### PR TITLE
React to Connection string config changes

### PR DESCRIPTION
**Pull request by HaoK**
cc @divega @danroth27 @phenning 

Let me know if you want me to just merge this

### CODE REVIEW

**Comment by rustd** [comment]
What's the motivation for this?

**Comment by HaoK** [comment]
https://github.com/aspnet/Configuration/issues/376

**Comment by rustd** [comment]
@sayedihashimi FYI in case this affects publishing to Azure App Service
@HaoK you would also need to update Startup where the connectionstring is accessed by EF

**Comment by HaoK** [comment]
Good catch on startup, I only grep'd json, my bad... Updated

File: src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json [hunk 1]
```diff
@@ -2,11 +2,7 @@
   "ApplicationInsights": {
     "InstrumentationKey": ""
   },
-  "Data": {
-    "DefaultConnection": {
-      "ConnectionString": "Server=(localdb)\\mssqllocaldb;Database=aspnet-$safeprojectname$-$guid2$;Trusted_Connection=True;MultipleActiveResultSets=true"
-    }
-  },
+  "DefaultConnection": "Server=(localdb)\\mssqllocaldb;Database=aspnet-$safeprojectname$-$guid2$;Trusted_Connection=True;MultipleActiveResultSets=true",
```
Comment (rustd) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: Can we keep the data section? It was something we had agreed upon earlier when reviewing templates. How  about something like the following?

``` C#
"Data" :{ "WebAppNameConnectionString":"<value>"}
or
"Data" :{ "ConnectionString":"<value>"}
```

Comment (divega) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: I understand that we had agreed previously on having the "Data" section but I think it means much less right now. 

The change that just went in Configuration is about removing the "Data:MyConnection:ConnectionString" convention we used to have and map to Azure Web Site connection strings. Now connection strings in Azure Web Sites will map directly to settings with the same name. See https://github.com/aspnet/Configuration/issues/376 for more context. This means we are no longer mangling the user provided name and gives users more freedom to structure their application settings in whatever way they want.

This also means that for customers to match "Data:ConnectionString" they will have to type exactly "Data:ConnectionString" in the Azure Web Site UI.

Personally I don't see much value in keeping the "Data" section in the default configuration in the template given that it sets precedent of a convention that is actually superfluous and unnecessary, but in theory we could keep it.

cc @Eilon @danroth27 What do you guys think? Anyone else needs to review this?

Comment (divega) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: cc @DamianEdwards in case he wants to weigh in.

**Comment by sayedihashimi** [comment]
Adding @bradygaster

File: src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json (hunk 1 above)
Comment (shirhatti) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: @rustd Weighing in.
The name you enter in the portal should be the name you use to access the connection string. 

After aspnet/Configuration#376, the only way to accommodate the `Data` section in the templates is have customers type `Data:MyConnection:ConnectionString` into the portal, which is just **awful**.

Unless I am missing context from why previously agreed to the `data` section, we should merge these changes in.

Comment (danroth27) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: Why not simply call the key ConnectionString instead of DefaultConnection? What does "default" even mean in this context?

Comment (bradygaster-zz) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: "DefaultConnection" is the connection string name developers have come to expect from the FNP experience. We also added UI to the SQL server/db-creation process in SDK 2.8.2 to re-surface the connection string per customer requests to make the experience more like the pre-2.8.1 experience. Users _expect_ to see the name of their DB connection string as DefaultConnection

Comment (bradygaster-zz) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: Also, agreed with @shirhatti on the verbosity of the connection string.It isn't pretty to have to type all that - making any assumptions about the nomenclature users "should use" for their connection string is kind of a bad proposition. I'm also chatting with the App Service folks to make sure there aren't any assumptions made on the names of the connection strings. We should make as few assumptions as possible about prefixes and so on.

Comment (danroth27) [inline src/Rules/StarterWeb/AI/IndividualAuth/appsettings.json]: Makes sense - DefaultConnection it is then.

**Comment by phenning** [comment]
Looks good to me, I agree with the more user friendly naming of DefaultConnection for Azure customers.

**Comment by phenning** [comment]
Are we OK taking this now?

**Comment by bradygaster-zz** [comment]
Adding @mlorbetske to the conversation. I'm okay with it but we're in verifying there won't be any negative impacts with the App Service team. Will want to discuss with Mike before we close.

**Comment by divega** [comment]
I would really like to help get this to closure, but I am having second thoughts on the whole change :smile:

Looking at the current default appsettings.json file I realized that we have a few sections already like `Logging` and `AppInsights`. Rather than having `DefaultConnection` appear as a top level key I am starting to like the idea of renaming `Data` to `ConnectionStrings` and (still) remove the unnecessary extra `:ConnectionString` level. E.g. in this file, it would look like this:

``` JSON
 "ConnectionStrings": {  
    "DefaultConnection": "Server=(localdb)\\mssqllocaldb;Database=aspnet-$safeprojectname$-$guid2$;Trusted_Connection=True;MultipleActiveResultSets=true"  
  },  
```

Besides having all the connection strings together in the JSON file another aspect I like about this is that we could eventually add bespoke extension methods for `IConfigurationRoot` to retrieve these well known sections in code, e.g.

``` C#
var defaultConnection = config.GetConnectionStrings()["DefaultConnection"];
```

I think it was @davidfowl who suggested we added something like that some time ago, and I guess I am finally coming to terms with it :smile:

**Comment by HaoK** [comment]
Yay for `config.GetConnectionString("DefaultConnection")`

**Comment by rustd** [comment]
I like the change to appsettings to have a section for connectionStrings. It could be used to add more connections to Storage and other data sources.
I am neutral on adding an API to get the connectionstring value.
+1 on taking in this change

**Comment by bradygaster-zz** [comment]
+1

**Comment by divega** [comment]
> Yay for  config.GetConnectionString("DefaultConnection") 

I think you have misread what I wrote :smile: (I actually missed the parenthesis, my bad).

**Comment by divega** [comment]
`GetConnectionStrings()` would be just sugar for `GetSection("ConnectionStrings")`.

**Comment by divega** [comment]
If people generally agree (I already talked to @rowanmiller and he is ok with it, I would like to hear from @danroth27 and others). Then we need to make the runtime change first.

**Comment by danroth27** [comment]
@divega @bradygaster Doesn't this effect publishing to Azure? I thought Azure publishing specifically creates a "DefaultConnection" environment variable.

**Comment by divega** [comment]
@danroth27 the new proposal doesn't change how the environment variable's name (it is some provider prefix + connection string name). But maybe I am missing something.

**Comment by bradygaster-zz** [comment]
@danroth27 @divega @mlorbetske tooling does default the connection string name being dropped in as "DefaultConnection." If tooling needs to change to support this we can change, but at the end of the day, we need to set the connection string name the customer _wants_ and not default it. our default was just to unbreak users who had issues with the deprecation of the automatic setting in SDK 2.8.1. We added DefaultConnection back in 2.8.2

**Comment by danroth27** [comment]
Let's not break the publish tooling unless we have a solid plan on how the team can continue to test publish to azure scenarios and how it will get fixed for customers.

**Comment by bradygaster-zz** [comment]
@danroth27 to clarify (just tested with the current bits). Whatever the customer provides for the name of the Connection String on the SQL provisioning dialog during App Service creation, is whatever gets dropped in as the name of the connection string in Azure. So if a user has this in their app settings JSON file (given the RC1 tooling and Azure SDK):

```
"Data": {
    "PersonConnection": {
      "ConnectionString": "Server=(localdb)\\mssqllocaldb;Database=PeopleDb;Trusted_Connection=True;MultipleActiveResultSets=true"
    }
  },
```

The name of the connection string they'll get when they publish their app to Azure is "PersonConnection." They'd then use code like this in their C# to get the connection string in the Startup.cs file. 

```
var connection = Configuration["Data:PersonConnection:ConnectionString"];
            services.AddEntityFramework()
                .AddSqlServer()
                .AddDbContext<PersonContext>(options => options.UseSqlServer(connection));
```

Looks like the templates need to be adjusted to support the shorter connection string name, but for Azure, we set whatever the user specifies.

**Comment by divega** [comment]
@bradygaster Are you saying that a configuration key called "Data:PersonConnection:ConnectionString" gets translated into a connection string called "PersonConnection"? Then it seems that the previous rules we had here in the runtime in RC1 are already hardcoded into the tooling and our hands are tied. Let's meet offline to sort this out.

**Comment by bradygaster-zz** [comment]
That's correct, @divega. Also verified with @davidebbo that the prefixing App Service is doing shouldn't impact this. According to @davidebbo "a connection string is a connection string" and the prefix nomenclature they're implementing shouldn't impact this story.

**Comment by HaoK** [comment]
Updated with proposal from https://github.com/aspnet/Configuration/issues/406

Looks a bit wordy to me, perhaps DefaultConnection => Default would be better given the nesting

**Comment by rustd** [comment]
@HaoK @divega What is the current state of this PR? Has this PR been updated for the final design?

**Comment by HaoK** [comment]
The current proposed design is what's in the PR I believe. To switch to using config.GetConnectionString instead of directly accessing config. If everyone is happy I can make the runtime changes.

**Comment by rustd** [comment]
I am fine with the change as long as this works for publishing to App Service. Default is better than DefaultConnection @divega you good with this?

**Comment by HaoK** [comment]
The runtime changes were checked in to Configuration with 9230cdb59ad38b74435bc39c0c8699d6a07c5e39

**Comment by rustd** [comment]
Thanks! Can you please update this PR to match the runtime changes.

**Comment by HaoK** [comment]
There's nothing to update

**Comment by rustd** [comment]
@phenning This can be merged with the next successful coherence signed build. Please make sure that when you publish the Web App to Azure you can register/ login
